src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -0,0 +1,1297 @@
+         * are not any weaker than the filters in the subplan. If a recheck filter is present, we can freely reduce
+         * selectivity of the subplan by e.g. removing intersection nodes, and we still get exactly same number of rows
+         * in the result set.
+         * </p>
+         * @param description textual description of the filter, for diagnostic purposes
+         * @param source source plan providing the input rows
+         * @param targetSelectivity a value in range [0.0, 1.0], but not greater than the selectivity of source
+         */
+        public RowsIteration filter(@Nonnull String description, @Nonnull RowsIteration source, double targetSelectivity)
+        {
+            RowsIterationCost sourceCost = source.cost();
+            Preconditions.checkArgument(targetSelectivity >= 0.0, "selectivity must not be negative");
+            Preconditions.checkArgument(targetSelectivity <= sourceCost.selectivity, "selectivity must not exceed source selectivity of " + sourceCost.selectivity);
+            return new Filter(this, nextId++, description, source, targetSelectivity);
+        }
+
+        /**
+         * Constructs a plan node that fetches only a limited number of rows.
+         * It is likely going to have lower fullCost than the fullCost of its input.
+         */
+        public RowsIteration limit(RowsIteration source, long limit)
+        {
+            return new Limit(this, nextId++, source, limit);
+        }
+    }
+
+
+    public static class CostCoefficients
+    {
+
+        /** Cost to advance the index iterator to the next key and load the key */
+        public final double unitCostPerKey;
+
+        /**
+         * Cost to advance the index iterator skipping to the desired point
+         * (includes the cost of converting between row ids and PrimaryKey back and forth)
+         */
+        public final double unitCostPerSkip;
+
+        /** Cost to get a key from DiskANN */
+        public final double unitCostPerScoredKey;
+
+        /**
+         * Cost to fetch one row from storage
+         */
+        public final double unitCostPerRow;
+
+
+        public CostCoefficients(double unitCostPerKey, double unitCostPerSkip, double unitCostPerScoredKey, double unitCostPerRow)
+        {
+            this.unitCostPerKey = unitCostPerKey;
+            this.unitCostPerSkip = unitCostPerSkip;
+            this.unitCostPerScoredKey = unitCostPerScoredKey;
+            this.unitCostPerRow = unitCostPerRow;
+        }
+    }
+
+    /** Convenience builder for building intersection and union nodes */
+    public static class Builder
+    {
+        final Factory factory;
+        final Operation.OperationType type;
+        final List<KeysIteration> subplans;
+
+        Builder(Factory context, Operation.OperationType type)
+        {
+            this.factory = context;
+            this.type = type;
+            this.subplans = new ArrayList<>(4);
```

Comment:
why are we pre-sizing the list ?

Reply:
I am slightly on the fence with that. Probably doesn't matter. My thinking was: If we have an intersection or union somewhere, then likely there are at least 2 children, maybe 3, but I almost never saw more than 4. Overallocating a bit just to cover the most frequent cases looks like a good compromise to me.